internal/pkg/service/buffer/condition/checker.go

```diff
@@ -21,64 +28,97 @@ const (
 	MinimalCredentialsExpiration = time.Hour
 )
 
-type checker struct {
-	*Service
-	logger           log.Logger
-	uploadConditions model.Conditions
+type Checker struct {
+	config      config.WorkerConfig
+	clock       clock.Clock
+	logger      log.Logger
+	client      *etcd.Client
+	schema      *schema.Schema
+	fileManager *file.Manager
+	dist        *distribution.Node
+	cachedStats *statistics.L1CacheProvider
 
 	checkLock    *sync.Mutex
 	exports      *etcdop.Mirror[model.ExportBase, cachedExport]
+	tokens       *etcdop.Mirror[model.Token, string]
 	activeSlices *etcdop.Mirror[model.Slice, cachedSlice]
 }
 
 type cachedExport struct {
 	key.ExportKey
-	ImportConditions model.Conditions
+	ImportConditions Conditions
 }
 
 type cachedSlice struct {
 	key.SliceKey
 	CredExpiration time.Time
 }
 
-func startChecker(s *Service) <-chan error {
-	c := &checker{
-		Service:          s,
-		logger:           s.logger.AddPrefix("[conditions]"),
-		uploadConditions: model.Conditions(s.config.UploadConditions),
-		checkLock:        &sync.Mutex{},
+type dependencies interface {
+	WorkerConfig() config.WorkerConfig
+	Clock() clock.Clock
+	Logger() log.Logger
+	Process() *servicectx.Process
+	EtcdClient() *etcd.Client
+	Schema() *schema.Schema
+	FileManager() *file.Manager
+	DistributionNode() *distribution.Node
+	StatisticsL1Cache() *statistics.L1CacheProvider
+}
+
+func NewChecker(d dependencies) (errCh <-chan error) {
+	c := &Checker{
+		config:      d.WorkerConfig(),
+		clock:       d.Clock(),
+		logger:      d.Logger().AddPrefix("[conditions]"),
+		client:      d.EtcdClient(),
+		schema:      d.Schema(),
+		fileManager: d.FileManager(),
+		dist:        d.DistributionNode(),
+		cachedStats: d.StatisticsL1Cache(),
+		checkLock:   &sync.Mutex{},
 	}
 
+	// Graceful shutdown
+	ctx, cancel := context.WithCancel(context.Background())
+	wg := &sync.WaitGroup{}
+	d.Process().OnShutdown(func() {
+		c.logger.Info("received shutdown request")
+		cancel()
+		wg.Wait()
+		c.logger.Info("shutdown done")
+	})
+
 	// Start watchers and ticker
 	initDone := make(chan error, 1)
-	c.Service.wg.Add(1)
+	wg.Add(1)
 	go func() {
-		defer c.Service.wg.Done()
+		defer wg.Done()
 		defer close(initDone)
 		startTime := c.clock.Now()
 
-		if err := <-c.watchExports(c.Service.ctx, c.Service.wg); err != nil {
+		if err := <-c.watchExports(ctx, wg); err != nil {
+			initDone <- err
+			return
+		}
+
+		if err := <-c.watchTokens(ctx, wg); err != nil {
```

Comment:
Pridane cachovanie `tokens`, ktore sa predtym pri kazdej operacii tahali z db.
Je to vyriesene uplne rovnako ako `watchExports` a `watchActiveSlices`